### PR TITLE
WP-946 Selectors

### PR DESCRIPTION
**Pull request by sergeychernyshev**
Prototype of selectors logic with support for <s>all 3</s> 2 ways of defining selectors:

1. individual `elementSelector` on each element entry (either a `string` for CSS selector or a `function`)
2. `elementSelector` configuration setting passed to `UXCapture.create()` accepts element as an argument
3. <s>`markSelector` configuration setting passed to `UXCapture.create()` accepts mark name as an argument (does not require zone configuration to have element abstraction layer)</s>

Also includes configuration example pages with micro SPA implementation.

JIRA: https://meetup.atlassian.net/browse/WP-946

### CODE REVIEW

File: src/Zone.js [hunk 1]
```diff
@@ -18,20 +20,77 @@ export default class Zone extends UXBase {
...
-			this.props.onMark(markName);
-			if (this.marks.every(m => m.marked)) {
-				this.measure(markName);
+		// mark names for elements already on the page
+		let alreadyOnThePage = [];
+
+		if (this.props.startMarkName === INTERACTIVE_TRANSITION_START_MARK_NAME) {
+			// if mark selector is configured, use it to find marks to record
+			if (props.markSelector) {
+				alreadyOnThePage.concat(
```
Comment (mmcgahan) [inline src/Zone.js]: `.concat` doesn't mutate the array. Consider

```js
alreadyOnThePage.push.apply(alreadyOnThePage, configuredMarkNames.filter(...))
```

in order to add elements to the existing array

File: examples/micro_spa.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+	setTimeout(() => {
+		// sets new view expectations
+		UXCapture.startView([
+			{
+				name: 'ux-destination-verified',
+				elements: [
+					{
+						marks: ['ux-text-title'],
+						// CSS selector, same as () => document.querySelectorAll('h1')
```
Comment (mmcgahan) [inline examples/micro_spa.js:29]: might want to always make this a function to keep the library code leaner and the interface more consistent

File: examples/selectors-elementtiming.html [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+		<script>
+			// sample custom reporter function for collected measures
+			const customMeasureRecordingFunction = name => {
+				// there can be multiple entries with the same name, get the latest
+				const measure = performance
+					.getEntriesByType('measure')
+					.filter(entry => entry.name === name)
+					.pop();
+
+				if (typeof measure.duration !== 'undefined') {
```
Comment (mmcgahan) [inline examples/selectors-elementtiming.html]: can measures ever have an undefined duration? Should this just be checking `if (measure) {`?

File: src/UXCapture.js [hunk 4]
```diff
@@ -45,6 +47,8 @@ const UXCapture = {
 	create: config => {
 		_onMark = config.onMark || NOOP;
 		_onMeasure = config.onMeasure || NOOP;
+		_markSelector = config.markSelector;
+		_elementSelector = config.elementSelector;
```
Comment (mmcgahan) [inline src/UXCapture.js:49]: I'm curious why we would want to support global mark/element selectors - at first look, it seems like it complicates the behavior quite a bit - are these global selectors implicit 'expected' requirements for every view? I'd like to force the client to be as explicit as possible for every view to avoid unexpected behavior, which would also make the code cleaner by only dealing with `markSelector` and `elementSelector` closer to where marks are handled in `ExpectedMark`

File: src/UXCapture.js [hunk 5]
```diff
@@ -45,6 +47,8 @@ const UXCapture = {
 	create: config => {
 		_onMark = config.onMark || NOOP;
 		_onMeasure = config.onMeasure || NOOP;
+		_markSelector = config.markSelector;
```
Comment (mmcgahan) [inline src/UXCapture.js]: What's the difference between `markSelector` and `elementSelector`? They look very similar in practice, and I'm wondering if we need both.

File: src/Zone.js [hunk 6]
```diff
@@ -18,20 +20,77 @@ export default class Zone extends UXBase {
...
+					configuredMarkNames.filter(
+						markName => props.markSelector(markName).length > 0
+					)
+				);
 			}
+
+			// if elements have selectors defined or global element selector is configured,
+			// use them to find marks to record
+			props.elements.forEach(element => {
```
Comment (mmcgahan) [inline src/Zone.js]: instead of relying on side effects to build the `alreadyOnThePage`, consider using `.reduce` to make this a little easier to read and maintain, e.g.

```js
const alreadyOnThePage = props.elements.reduce((acc, element) => {
  const elementSelector = element.elementSelector || this.props.elementSelector;
  const selected = typeof elementSelector === 'function'
    ? elementSelector()
    : document.querySelectorAll(elementSelector);
  if (selected.length > 0) {
    return acc.concat(element.marks);
  }
  return acc;
}, []);
```

**Review by mmcgahan — COMMENTED** [review]
This example seems to be supporting a number of overlapping interfaces, which makes it a little hard to follow - it looks like `elementSelector` can be defined at the global level or by Zone, and can be a CSS selector string or a function. `markSelector` is also a little hard to distinguish from using a function as an `elementSelector`. Can we narrow it down to be a bit more prescriptive on the first pass?

We should continue aiming to keep the library as lean as possible since it's expected to block page render while loading.

File: src/Zone.js (hunk 1 above)
Comment (sergeychernyshev) [reply src/Zone.js]: Yeah, or do array spread. I keep hitting it, especially when I refactor code which just worked ;)

File: examples/selectors-elementtiming.html (hunk 3 above)
Comment (sergeychernyshev) [reply examples/selectors-elementtiming.html]: I don't remember why I made it so elaborate for the sample pages. I assume it should work with just checking if measure exists, will try it out.

File: src/UXCapture.js (hunk 4 above)
Comment (sergeychernyshev) [reply src/UXCapture.js:49]: We don't have to support all 3 interfaces, I just wanted to add solution that works with all the examples we discussed.

In reality, I believe markSelector is just a hack and has no good reason to exist except for being easiest to add it to existing configs used with current version of UX Capture.

And elementSelector to be both global (e.g. established rule for entire site) and defined for individual element (more flexibility, less centralization of solution) is a good idea. Both can be implemented by React client.

File: src/UXCapture.js (hunk 5 above)
Comment (sergeychernyshev) [reply src/UXCapture.js]: Agreed, see above.

File: src/UXCapture.js (hunk 4 above)
Comment (mmcgahan) [reply src/UXCapture.js:49]: I think my main concern with an `elementSelector` global is that it means that each view instance isn't fully in control of defining its expected marks, which leads to ambiguous behavior, e.g. "Why isn't my 'primary content loaded' measure firing even though all of the marks I passed into the view configuration have fired?!?"

If we can always say "the measures for a particular view always correspond to the marks passed into `startView`", the tool will be easier to use and understand.

File: src/Zone.js (hunk 1 above)
Comment (mmcgahan) [reply src/Zone.js]: tangent, but array spread ends up adding significant noise to the transpiled output script, so I think we should avoid it in this lib for now

File: examples/micro_spa.js (hunk 2 above)
Comment (sergeychernyshev) [reply examples/micro_spa.js:29]: I'd show two different interfaces that are supported here, will make one of them as CSS selector and another a function.

File: src/UXCapture.js (hunk 4 above)
Comment (sergeychernyshev) [reply src/UXCapture.js:49]: Global selector function can rely on convention established by specific site.

For example in [element timing example](https://github.com/meetup/ux-capture/blob/bd294bb074f805b1b7117a29bda9091506706067/examples/selectors-elementtiming.html#L27) this convention is element's name set as `elementtiming` attribute's value on the element. Same goes for universal selector example which [uses class name](https://github.com/meetup/ux-capture/blob/bd294bb074f805b1b7117a29bda9091506706067/examples/selectors-universal.html#L27) as convention.

Because global selector takes element object from zone definition as an attribute it allows both convention and flexibility.

Comment (sergeychernyshev) [reply src/UXCapture.js:49]: BTW, in current implementation, individual elementSelectors override global one in case view wants to deviate from a convention.

**Comment by sergeychernyshev** [comment]
This PR is waiting for mark-history-based approach verdict.
